Feat: Slash command integration

aBitEmberous opened the pull request:
This PR will add Slash commands to the Framework. It also adds a decent chunk of In-Editor documentation to the new Slash Commands but also to Utils along with several other things.

aBitEmberous committed:
Started work on Slash Command integration

aBitEmberous committed:
feat: Added a validation step
If a command is a slash command but somehow is loaded as a regular, it will be cancelled and a warning is put out.

aBitEmberous committed:
feat: Slash Command implementation

Slash commands can now be run and given arguments.

aBitEmberous committed:
change: Changed the Slash command to now ask for a confirm

aBitEmberous committed:
docs: Added a ton of missing documentation.

aBitEmberous committed:
docs: Additional documentation
CommantEvent is now also documented and explains what is relevant to what.

aBitEmberous committed:
fix: Updated SlashCommandOption to be an interface instead of type

aBitEmberous committed:
docs: Updated readme with information on Slash Commands

aBitEmberous committed:
fix: Slash commands can now be removed from the app

Strikeeaglechase commented on src/command.ts:
```diff
@@ -29,12 +39,15 @@ abstract class Command {
 	altNames: string[] = [];
 	help: { msg?: string; usage?: string; } = {}
+	slashCommand?: boolean = false;
+	slashOptions?: SlashCommandOption[] = [];
```
Slash commands should be their own class that extends the Command class with these fields as required.

Strikeeaglechase commented on src/command.ts:
```diff
@@ -49,15 +62,25 @@ abstract class MultiCommand extends Command {
-	constructor(frameworkOrEvent: FrameworkClient, message: Discord.Message, app: T, command: BotCommand)
-	constructor(frameworkOrEvent: FrameworkClient | CommandEvent, message?: Discord.Message, app?: T, command?: BotCommand) {
+	constructor(frameworkOrEvent: FrameworkClient, message: Discord.Message, app: T, command: BotCommand, interaction?: Discord.CommandInteraction)
+	constructor(frameworkOrEvent: FrameworkClient | CommandEvent, message?: Discord.Message, app?: T, command?: BotCommand, interaction?: Discord.CommandInteraction) {
```
Keep normal constructor, move to seperate class

Strikeeaglechase commented on src/app.ts:
```diff
@@ -143,6 +185,28 @@ class FrameworkClient {
+		// Check to see if we are forced to reset all slash commands.
+		// This can be useful in the event a bad command got registered and needs to be removed.
+		if (this.options.slashCommandReset) {
+			await this.deleteSlashCommands();
```
fetchBotCommands is called multiple times over the lifecycle of the bot, running this here will cause problems

Strikeeaglechase commented on src/app.ts:
```diff
@@ -143,6 +185,28 @@ class FrameworkClient {
+
+					// Yes, I am doing this on this and yes, this is not a good idea. It just gotta work now.
+					this.slashCommands.push(element);
+					imported.splice(imported.indexOf(element), 1);
```
If there are two slash commands back to back, this will only load the first one then skip the second one, don't modify loops while reading them

Strikeeaglechase commented on src/app.ts:
```diff
@@ -224,6 +294,23 @@ class FrameworkClient {
+		let commandName = interaction.commandName.toLowerCase();
+		let command:Command = this.slashCommands.find(cmd => cmd.name.toLowerCase() == commandName);
+
+		if(!command) this.error("Somehow this command does not exist..");
```
this.error just creates an embed that can be used, it doesn't actually send the error anywhere. Need a log, send to a user, then the function should exit rather than continuing

Strikeeaglechase commented on src/test/commands/slash/slash.ts:
```diff
@@ -0,0 +1,46 @@
+    // Run when the command is called.
+    async run(event: CommandEvent) {
+        const {framework, interaction} = event;
+        if(!event.interaction.isCommand()) return; // required bc typescript funni
```
You should type the interaction to be specifically a command interaction so that this isn't required

aBitEmberous committed:
fix: Fixed an issue where commands were incorrectly aborted

Slash commands were incorrectly aborted if they did not exist. This has been resolved.

aBitEmberous committed:
fix: Moved where Slash command Reset is checked

Slash command reset is now checked in the Framework.Init function rather then in Framework.fetchBotCommands

aBitEmberous committed:
fix: Ported Slash Commands / Evebts to their own subclasses

Slash commands now use a subclass of Command instead of regular Command. Slash Events also have been ported.

aBitEmberous committed:
fixes: SlashCommand fixes in App.ts.

Fixed a bunch of reference issues regarding the original Slash in Command setup. Additionally, fixed an issue with Command loading where slash commands could be loaded mutliple times.

Fixed an issue related to slash command loading where it might skip a command.

aBitEmberous committed:
docs: Updated README.
